### PR TITLE
[AIRFLOW-4935] Add method in the bigquery hook to list tables in a dataset

### PR DESCRIPTION
**Pull request by benjamingrenier**
Method `get_dataset_tables_list` added in the hook, permits to use a table prefix too.

### CODE REVIEW

File: airflow/contrib/hooks/bigquery_hook.py [hunk 1]
```diff
@@ -1718,6 +1718,79 @@ def get_datasets_list(self, project_id=None):
 
         return datasets_list
 
+    def get_dataset_tables_list(self,dataset_id, project_id=None, table_prefix=None, max_results=None):
```
Comment (ryanyuan) [inline airflow/contrib/hooks/bigquery_hook.py]: space between `self,` and `dataset_id`

**Comment by ryanyuan** [comment]
Could you write some unit tests for it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 2]
```diff
@@ -1718,6 +1718,79 @@ def get_datasets_list(self, project_id=None):
...
+            For more information, see:
+            https://cloud.google.com/bigquery/docs/reference/rest/v2/tables/list
+
+        :param dataset_id: The BigQuery Dataset ID
+        :type dataset_id: str
+        :param project_id: The GCP Project ID
+        :type project_id: str
+        :param table_prefix: Tables must begin by this prefix to be returned
+        :type table_prefix: str
+        :param max_results:
```
Comment (ryanyuan) [inline airflow/contrib/hooks/bigquery_hook.py]: description of max_results missing

File: airflow/contrib/hooks/bigquery_hook.py [hunk 3]
```diff
@@ -1718,6 +1718,79 @@ def get_datasets_list(self, project_id=None):
...
+        dataset_project_id = project_id if project_id else self.project_id
+
+        optional_params = {}
+        if max_results:
+            optional_params['maxResults'] = max_results
+
+        dataset_tables_list = []
+        next_page_token = None
+        while True:
+            if next_page_token:
```
Comment (ryanyuan) [inline airflow/contrib/hooks/bigquery_hook.py]: This if check can be gone since next_page_token will be checked later:
`
 if not next_page_token:
`

**Comment by zzlbuaa** [comment]
Hi, this hook method is a useful feature for our use case. I'm wondering are you still working on this PR? 
cc @criccomini @whynick1

**Comment by benjamingrenier** [comment]
Yes I'm making some unit tests. It will be available soon.

File: airflow/contrib/hooks/bigquery_hook.py [hunk 4]
```diff
@@ -1718,6 +1718,79 @@ def get_datasets_list(self, project_id=None):
...
+                optional_params['PageToken'] = next_page_token
+
+            try:
+                tables_list_resp = self.service.tables().list(
+                    projectId=dataset_project_id,
+                    datasetId=dataset_id,
+                    **optional_params).execute()
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: Can you add a retry here?

**Comment by zzlbuaa** [comment]
@benjamingrenier any updates on this?

**Comment by benjamingrenier** [comment]
I will finish it this week.

Le mar. 6 août 2019 à 00:01, Zhengliang(Zane) <notifications@github.com> a
écrit :

**Comment by mik-laj** [comment]
Are you planning to continue working on this change? This week, I would like to deal with PR reviews related to GCP. I would be happy if you would respond to all comments

**Comment by benjamingrenier** [comment]
Yes I will work on it this week. I try to resolve my docker errors to run unit-tests locally.

File: airflow/contrib/hooks/bigquery_hook.py (hunk 1 above)
Comment (benjamingrenier) [reply airflow/contrib/hooks/bigquery_hook.py]: Thanks

File: airflow/contrib/hooks/bigquery_hook.py (hunk 2 above)
Comment (benjamingrenier) [reply airflow/contrib/hooks/bigquery_hook.py]: Thanks

File: airflow/contrib/hooks/bigquery_hook.py (hunk 3 above)
Comment (benjamingrenier) [reply airflow/contrib/hooks/bigquery_hook.py]: Done ! Thanks.

File: airflow/contrib/hooks/bigquery_hook.py (hunk 4 above)
Comment (benjamingrenier) [reply airflow/contrib/hooks/bigquery_hook.py]: Yes, but why ?

Comment (mik-laj) [reply airflow/contrib/hooks/bigquery_hook.py]: Good practice is to repeat requests in error response.
Please run following command for example
`grep -r execute airflow/`
More information is also available in the GCP's official integration guide

**Comment by mik-laj** [comment]
Recently introduced a new development environment.  have you tried them? we plan to create a new extension to improve GCP development.

File: airflow/contrib/hooks/bigquery_hook.py [hunk 5]
```diff
@@ -1718,6 +1717,75 @@ def get_datasets_list(self, project_id=None):
 
         return datasets_list
 
+    def get_dataset_tables_list(self, dataset_id, project_id=None, table_prefix=None, max_results=None):
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: What do you think about adding the @GoogleCloudBaseHook.fallback_to_default_project_id``decorator? This will allow you to delete the repeating code, but also add validation - this field is required.

Reference: https://github.com/apache/airflow/pull/5546/files#r307964979

File: airflow/contrib/hooks/bigquery_hook.py [hunk 6]
```diff
@@ -492,7 +492,6 @@ def create_external_table(self,  # pylint: disable=too-many-locals,too-many-argu
         }
 
         if source_format in src_fmt_to_param_mapping.keys():
-
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 7]
```diff
@@ -844,8 +843,8 @@ def run_query(self,  # pylint: disable=too-many-locals,too-many-arguments
                 'createDisposition': create_disposition,
             })
 
-        if 'useLegacySql' in configuration['query'] and configuration['query']['useLegacySql'] and\
-                'queryParameters' in configuration['query']:
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 8]
```diff
@@ -857,14 +856,14 @@ def run_query(self,  # pylint: disable=too-many-locals,too-many-arguments
...
     def run_extract(  # noqa
-            self,
-            source_project_dataset_table,
-            destination_cloud_storage_uris,
-            compression='NONE',
-            export_format='CSV',
-            field_delimiter=',',
-            print_header=True,
-            labels=None):
+        self,
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 9]
```diff
@@ -957,9 +956,9 @@ def run_copy(self,  # pylint: disable=invalid-name
         :type labels: dict
         """
         source_project_dataset_tables = ([
-            source_project_dataset_tables
-        ] if not isinstance(source_project_dataset_tables, list) else
-            source_project_dataset_tables)
+                                             source_project_dataset_tables
+                                         ] if not isinstance(source_project_dataset_tables, list) else
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 10]
```diff
@@ -969,11 +968,11 @@ def run_copy(self,  # pylint: disable=invalid-name
                                  var_name='source_project_dataset_table')
             source_project_dataset_tables_fixup.append({
                 'projectId':
-                source_project,
+                    source_project,
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 11]
```diff
@@ -1118,11 +1117,11 @@ def run_load(self,  # pylint: disable=too-many-locals,too-many-arguments,invalid
             'ALLOW_FIELD_ADDITION', "ALLOW_FIELD_RELAXATION"
         ]
         if not set(allowed_schema_update_options).issuperset(
-                set(schema_update_options)):
+            set(schema_update_options)):
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 12]
```diff
@@ -1118,11 +1117,11 @@ def run_load(self,  # pylint: disable=too-many-locals,too-many-arguments,invalid
...
             raise ValueError(
                 "{0} contains invalid schema update options."
                 "Please only use one or more of the following options: {1}"
-                .format(schema_update_options, allowed_schema_update_options))
+                    .format(schema_update_options, allowed_schema_update_options))
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 13]
```diff
@@ -1253,7 +1252,7 @@ def run_with_configuration(self, configuration):
                 else:
                     raise Exception(
                         'BigQuery job status check failed. Final error was: {}'.
-                        format(err.resp.status))
+                            format(err.resp.status))
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 14]
```diff
@@ -1308,7 +1307,7 @@ def poll_job_complete(self, job_id):
             else:
                 raise Exception(
                     'BigQuery job status check failed. Final error was: {}'.
-                    format(err.resp.status))
+                        format(err.resp.status))
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 15]
```diff
@@ -1317,7 +1316,7 @@ def cancel_query(self):
         """
         jobs = self.service.jobs()
         if (self.running_job_id and
-                not self.poll_job_complete(self.running_job_id)):
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 16]
```diff
@@ -1468,10 +1467,10 @@ def run_table_upsert(self, dataset_id, table_resource, project_id=None):
                         body=table_resource).execute(num_retries=self.num_retries)
             # If there is a next page, we need to check the next page.
             if 'nextPageToken' in tables_list_resp:
-                tables_list_resp = self.service.tables()\
+                tables_list_resp = self.service.tables() \
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 17]
```diff
@@ -1468,10 +1467,10 @@ def run_table_upsert(self, dataset_id, table_resource, project_id=None):
...
                     .list(projectId=project_id,
                           datasetId=dataset_id,
-                          pageToken=tables_list_resp['nextPageToken'])\
+                          pageToken=tables_list_resp['nextPageToken']) \
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 18]
```diff
@@ -2004,7 +2072,6 @@ def _bq_cast(string_field, bq_type):
 
 
 def _split_tablename(table_input, default_project_id, var_name=None):
-
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 19]
```diff
@@ -2022,7 +2089,7 @@ def var_print(var_name):
     if table_input.count('.') + table_input.count(':') > 3:
         raise Exception(('{var}Use either : or . to specify project '
                          'got {input}').format(
-                             var=var_print(var_name), input=table_input))
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: airflow/contrib/hooks/bigquery_hook.py [hunk 20]
```diff
@@ -2036,7 +2103,7 @@ def var_print(var_name):
     else:
         raise Exception(('{var}Expect format of (<project:)<dataset>.<table>, '
                          'got {input}').format(
-                             var=var_print(var_name), input=table_input))
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

File: tests/contrib/hooks/test_bigquery_hook.py [hunk 21]
```diff
@@ -703,6 +703,50 @@ def test_get_datasets_list(self):
                 project_id=project_id)
             self.assertEqual(result, expected_result['datasets'])
 
+    def test_get_dataset_tables_list(self):
```
Comment (mik-laj) [inline tests/contrib/hooks/test_bigquery_hook.py]: It's look like unrelated changes. Can you revert it?

**Comment by mik-laj** [comment]
Can you do rebase, please?

**Comment by benjamingrenier** [comment]
> Can you do rebase, please?

It's done. Ok for you ?

**Comment by mik-laj** [comment]
Something is wrong with this PR.  Too many commits.

I'm on vacation now🇫🇷. I'll look at this on Friday.

**Comment by benjamingrenier** [comment]
On which branch do you want the rebase ?

Le mar. 20 août 2019 à 10:23, Kamil Breguła <notifications@github.com> a
écrit :

**Comment by ryanyuan** [comment]
@benjamingrenier After rebasing, squash your commits and then do a force push.

**Comment by mik-laj** [comment]
I suspect that the references to the main branch are broken and the history will have to be rewritten (git reset --hard, git cherry-pick).

**Comment by benjamingrenier** [comment]
I think it's all right.

File: airflow/contrib/hooks/bigquery_hook.py [hunk 22]
```diff
@@ -1793,6 +1793,68 @@ def get_datasets_list(self, project_id=None):
...
+
+            for table in response.get('tables', []):
+                table_ref = table.get('tableReference')
+                table_id = table_ref.get('tableId')
+                if table_id and (not table_prefix or table_id.startswith(table_prefix)):
+                    dataset_tables_list.append(table_ref)
+
+            request = self.service.tables().list_next(previous_request=request,
+                                                      previous_response=response)
+        self.log.info(dataset_tables_list)
```
Comment (mik-laj) [inline airflow/contrib/hooks/bigquery_hook.py]:
```suggestion
```
I haven't seen it before, but it seems to me that there will be too much information in the logs.

Comment (benjamingrenier) [reply airflow/contrib/hooks/bigquery_hook.py]: I will fix it soon. I prepare a commit with more unit tests too.

**Comment by mik-laj** [comment]
Hi.

I made a change in the base class - GoogleCloudBaseHook. Your PR may need to be changed. Could you do rebase?

Cheers

Reference:
https://github.com/apache/airflow/pull/5907

File: airflow/contrib/hooks/bigquery_hook.py (hunk 22 above)
Comment (potiuk) [reply airflow/contrib/hooks/bigquery_hook.py]: Can you please remove the log before merge?

**Review by potiuk — CHANGES_REQUESTED** [review]
@benjamingrenier We are in the middle of moving hooks and operators from contrib to core. Is it possible to rebase (add more unittestss?) quickly so that we can merge the change before we move it?

**Comment by benjamingrenier** [comment]
> @benjamingrenier We are in the middle of moving hooks and operators from contrib to core. Is it possible to rebase (add more unittestss?) quickly so that we can merge the change before we move it?

Rebase done and unittests added

**Comment by potiuk** [comment]
Hey @benjamingrenier -> I must ask for another rebase: (.

**Comment by benjamingrenier** [comment]
OK, it will be done in some hours.

Le sam. 21 sept. 2019 à 21:02, Jarek Potiuk <notifications@github.com> a
écrit :

**Comment by whynick1** [comment]
I am also interested in this PR, any update? 😄
@potiuk @mik-laj @ryanyuan

**Comment by benjamingrenier** [comment]
Someone can do the review ?

**Comment by mik-laj** [comment]
I added this PR to the queue next week. I will try to review it as soon as possible.

**Review by mik-laj — APPROVED** [review]
